KAFKA-4848: Fix retryWithBackoff deadlock issue

sjmittal opened the pull request:
Fixes related to handling of MAX_POLL_INTERVAL_MS_CONFIG during deadlock and CommitFailedException on partition revoked.

sjmittal committed:
added git ignores

added git ignores

sjmittal committed:
Fix for backoffTimeMs and CommitFailedException

Fix for backoffTimeMs in AbstractTaskCreator -> retryWithBackoff and
CommitFailedException in suspendTasksAndState -> commitOffsets

sjmittal committed:
undo change as not valid in trunk

due to change that now  CommitFailedException isn't thrown anymore there
is no need to handle it in catch part.

guozhangwang commented on clients/.gitignore:
```diff
@@ -0,0 +1 @@
+/bin/
```
I think this makes sense, but generally speaking we'd tend to have one "fix" per PR for easier reviews. Ditto below.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -288,6 +291,8 @@ public StreamThread(TopologyBuilder builder,
 
         this.stateDirectory = new StateDirectory(applicationId, config.getString(StreamsConfig.STATE_DIR_CONFIG), time);
+        Object maxPollInterval = consumerConfigs.get(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG);
+        this.rebalanceTimeoutMs =  maxPollInterval == null ? 300000 : (Integer) ConfigDef.parseType(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, maxPollInterval, Type.INT);
```
Any rationale to have a magic number `300000` here?

guozhangwang replied:
I think we can just use `consumerConfigs.getInt(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG)` will automatically return its default value if it was not set.

sjmittal replied:
@guozhangwang How I understand is that default value of max.poll.interval.ms is 300000 and consumerConfigs map is built from StreamsConfig and if in those configs user has not specified this consumer config property then the value would be null and hence it should use the default one.
note that consumerConfigs are of type Map<String, Object> so we cannot use `consumerConfigs.getInt(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG)`.

guozhangwang replied:
In `StreamsConfig` we do populate the map from consumer / producer's default values first then use user specified values to overwrite, so it should be safe to `config.getInt` where `config` is of type `StreamsConfig`.

sjmittal replied:
Could you please point me to the code where we load all default consumerConfigs inside `StreamsConfig `. As far as I am able to see is that to consumer we pass any values that we override in streamsConfig. The defaults of `ConsumerConfig` are intenal to consumers only. That is why I had to copy the default value for `ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG` in case streamsConfig did not specify one.

guozhangwang replied:
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L453

Note that inside the function it calls `originals()` which is parsed on the consumer config in which default values will be auto-filled when it is not specified by the user.

sjmittal replied:
Well I checked that code but please check this line
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L550
Here `configNames` are consumer config names, however `originals()` is still defaulted by  `StreamsConfig#CONFIG ` as per line https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L217 and https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java#L449.
Now this `StreamsConfig#CONFIG ` does not have `MAX_POLL_INTERVAL_MS_CONFIG` defined by default, so if user does not provide this property in streams config, then it will never find in the `StreamsConfig#CONFIG ` map. Hence I had to check against the default value in `ConsumerConfig` if it is not provided.

dguy replied:
Rather than this i think it would be better to add:
`tempConsumerDefaultOverrides.put(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, "300000");`
to the `CONSUMER_DEFAULT_OVERRIDES` map in `StreamsConfig`

guozhangwang replied:
@sjmittal You are right, I misinterpreted the line here.

I agree with @dguy that we could add this into `CONSUMER_DEFAULT_OVERRIDES` inside `StreamsConfig`. Note that it will only take effect if users do NOT specify these configs themselves.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```
I think there is a small risk that if the lock was kept by someone else for more than rebalanceTimeout but it was released later, then this task will be "lost" as it is not picked by anyone. But I think it will be covered by KIP-129 anyways. cc @mjsax

mjsax replied:
Ack. Will have a detailed look into this when doing KIP-129

guozhangwang reviewed (commented):
Could you change the title to add `KAFKA-4848:` as the prefix?

guozhangwang commented:
@sjmittal Another thought about resolving this deadlock thing, is that we can sort the tasks (i.e. the `newTasks` generated in the `addStreamTasks` and `addStandbyTasks`) to be initialized by their task-ids, so that there will not be a scenario where one thread is grabbing lock for task1 then lock for task2 while the other thread is trying to grab lock for task2 then lock for task1, hence the deadlock.

Even if the deadlock involves more than two threads, having the tasks sorted will make sure that at least one pair of the threads will break the tie with each other and hence there will be no deadlocks. WDYT?

sjmittal commented:
@guozhangwang To some extent it makes sense when we talk in 2 threads scenario. However I am not able to visualize when there are more than 2 threads and tasks of thread 1 are moved to different threads and if we can have some three way deadlock.

guozhangwang commented:
@sjmittal I see. Yeah it seems with potential multi-way dead lock it won't work: For example, imagine thread1 owning task1/2, thread2 owning task 3/4, thread3 owning task 5/6, and lets assume thread1 is kicked and the new assignment is thread2 -> task 1/3/4 and thread3 -> task 2/5/6, assuming thread1 already grab lock on task1 before being kicked but not on task2, then thread2 trying to grab locks for task 1/3/4 will be blocked on 1, but thread3 grabbing locks on 2/5/6 happen to succeed on 2 and proceeds, we will leave with thread1 and thread2 deadlocked then.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```
I'm not sure this is correct. Lets say we have 2 tasks. Currently,  once the first task is successful we will sleep for `backoffTimeMs`. This shouldn't happen, i.e., once the task was successful `backoffTimeMs` should be set back to the initial value and we shouldn't sleep before we attempt to create the next task.

We need to track how long we've been rebalancing in total and use that to determine if we should give up (rather than the `backoffTimeMs`

In the event that it times out i think we should probably throw an exception so we leave the group, the thread shuts down (releasing any acquired locks), and some other thread can pick up the work.

sjmittal replied:
`MAX_POLL_INTERVAL_MS_CONFIG` is absolute max time till this thread is kicked out. So this can be the upper limit for `backoffTimeMs`. I am not sure how can we track total time this thread was re balancing.  Also not throwing an exception but carrying on is OK, because it would try to commit offsets and it will fail that time, because it is no longer the owner of these partitions and already kicked out of the group. So it will try to join again and get new partitions.

dguy replied:
Sorry - i'm reading it wrong. I was failing to see that the backoff is done in the outer loop. Anyway, i still think the `backoffTimeMs` should be reset after each successful task creation. 

As for tracking the total rebalance time... we could just get the time at the start of `onPartitionsAssigned` and make sure that `MAX_POLL_INTERVAL_MS` hasn't passed since then.

Yes you probably could rely on the `CommitFailedException` to be thrown by poll. My concern is that we carry on for some reason and some tasks are never processed. In which case it would be better to find a way to bail early so they can be re-assigned.

mjsax replied:
> Anyway, i still think the backoffTimeMs should be reset after each successful task creation.

I agree -- we missed this in the original implementation -- should have been like this from beginning on.

guozhangwang replied:
Thinking it a bit more, I feel even after KIP-129 this issue will still persists as long as we create one task and grab its local state file lock at a time. So how about the following solution (I feel it is not very elegant but could be correct in general):

1. In the `onPartitionAssigned`, do not create the tasks (both active and standby) but only does the cleaning of suspended ones; and remember the tasks needed to be created.

2. Then in the stream's main loop, check if there are any tasks that need to be created; if there are try creating them all by first grab all the locks and then creating them. If failed to grabbing all the locks then release the already-grabbed ones and the continue to the next loop, in which `consumer.poll()` will be triggered again and then it will be notified if it has lost the ownership or not. The key idea here is that either all tasks are created or none should be created.

3. after KIP-129 is introduced, we need to think whether the above pattern could be improved with `producer.initTxn` cals.

WDYT?

guozhangwang replied:
anyways this could be in a separate PR just throwing it out for discussions @sjmittal @dguy @mjsax We can merge the PR as is if other reviewers could give a thumb up.

mjsax replied:
Sounds good to me.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```
You can get the time when `onPartitionsAssigned` is called and pass that through. It may not be perfect, but i think it will be close enough. You then just compare if `MAX_POLL_INTERVAL_MS` has passed.
Anyway, `rebalanceTimeMs` needs to be reset whenever a task is successfully created. Lets say that `MAX_POLL_INTERVAL_MS` is set to 400 and i have 5 tasks to create. 
Currently what would happen:
1. First task created
2. Sleep for 50
3. Second task created
4. Sleep for 100
5. Third task created
6. Sleep for 200
7. Fourth task created
8. Sleep for 400
9. break

The point being that we've wasted almost double `MAX_POLL_INTERVAL_MS` by sleeping when we didn't need to. So by the time we get to 8 - it is probably too late anyway.

sjmittal committed:
KAFKA-4848: moved ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG setting

moved default ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG setting to
SteamsConfig.

sjmittal committed:
KAFKA-4848: reset backoffTimeMs and check against start of partion assignment

reset backoffTimeMs when task is created, also check against start of of
partion assignment instead of current value of backoffTimeMs

sjmittal commented:
@guozhangwang  @dguy  Could you please approve the latest changes, in case there are some changes to be done, please let me know.
Thanks!

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -288,6 +291,8 @@ public StreamThread(TopologyBuilder builder,
         this.standbyRecords = new HashMap<>();
 
         this.stateDirectory = new StateDirectory(applicationId, config.getString(StreamsConfig.STATE_DIR_CONFIG), time);
+        Object maxPollInterval = consumerConfigs.get(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG);
```
nit: `final`

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -288,6 +291,8 @@ public StreamThread(TopologyBuilder builder,
 
         this.stateDirectory = new StateDirectory(applicationId, config.getString(StreamsConfig.STATE_DIR_CONFIG), time);
+        Object maxPollInterval = consumerConfigs.get(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG);
+        this.rebalanceTimeoutMs =  (Integer) ConfigDef.parseType(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, maxPollInterval, Type.INT);
```
It is a shame we have to do it like this, but i don't see a viable alternative

dguy reviewed (approved):
1 nit pick, but otherwise LGTM

sjmittal commented:
@dguy  Please review and confirm the other changes too where I am resetting `backoffTimeMs` after task creation too and checking the time elapsed against start of partition assigned time.
cc: @guozhangwang

sjmittal committed:
KAFKA-4848: minor fix

minor fix

dguy reviewed (approved):
Thanks for the patch @sjmittal - LGTM

guozhangwang commented:
LGTM except the early question I had about `.gitingore` changes: are these intentional?

sjmittal commented:
Yes git ignore changes are intentional as doing development via eclipse it adds those directories so for a clean status check it is necessary to ignore those directories and files. If all is OK please merge the same.

guozhangwang commented:
> Yes git ignore changes are intentional as doing development via eclipse it adds those directories

Hmm I was thinking it was for different purposes, since I thought this issue is already fixed as in

https://github.com/apache/kafka/pull/2382/files
https://github.com/apache/kafka/pull/2351/files

And the above changes seem to be overlapping with these commits? cc @ijuma

sjmittal commented:
Well these changes will not overlap other PRs which worked on this area.
It will just help developers avoid seeing many un versioned files, especially when they are working with project specific eclipse settings created before applying the gradle eclipse fix.
@guozhangwang @ijuma

guozhangwang commented:
> when they are working with project specific eclipse settings created before applying the gradle eclipse fix.

Thanks for the explanation.

guozhangwang commented:
Merged to trunk.

guozhangwang commented:
Will also try to cherry-pick to 0.10.2 when 0.10.2.1 is released.
